### PR TITLE
Explain fully qualified syntax for `Rc` and `Arc`

### PR DESCRIPTION
**Pull request by camelid**
Also cleaned up some other small things.

@rustbot modify labels: T-doc

### CODE REVIEW

File: library/alloc/src/rc.rs [hunk 1]
```diff
@@ -35,15 +35,26 @@
 //! `Rc<T>` automatically dereferences to `T` (via the [`Deref`] trait),
 //! so you can call `T`'s methods on a value of type [`Rc<T>`][`Rc`]. To avoid name
 //! clashes with `T`'s methods, the methods of [`Rc<T>`][`Rc`] itself are associated
-//! functions, called using function-like syntax:
+//! functions, called using [fully qualified syntax]:
```
Comment (matklad) [inline library/alloc/src/rc.rs:38]: Do we actually call this "fully qualified syntax"? I thought that would be reserved for `<RC>::downgrade` form. 

Naively, "function like syntax" or "associated function syntax" seem like better options to me here.

File: library/alloc/src/rc.rs [hunk 2]
```diff
@@ -35,15 +35,26 @@
...
+//! `Rc<T>`'s implementations of traits like `Clone` should also be called using
+//! fully qualified syntax to avoid confusion as to whether the *reference* is being
+//! cloned or the *backing data* (`T`) is being cloned:
+//!
+//! ```
+//! use std::rc::Rc;
+//!
+//! let my_rc = Rc::new(());
+//! let your_rc = Rc::clone(&my_rc);
+//! ```
```
Comment (matklad) [inline library/alloc/src/rc.rs:56]: Note that we explicitly decided against this in https://github.com/rust-lang/rust/pull/63252.

My personal opinion is that people do use `Arc::clone` syntax in the wild, so this *should* be in the official docs one way or another, but probably without "should" or "idiomatic" langauge.

Comment (matklad) [reply library/alloc/src/rc.rs:56]: Urgh, but #63252 didn't update

https://github.com/rust-lang/rust/blame/master/library/alloc/src/rc.rs#L64-L66

I guess, the status quo is confused here, and we should probably try to be consistent at least withing stdlib itself :)

File: library/alloc/src/rc.rs (hunk 1 above)
Comment (steveklabnik) [reply library/alloc/src/rc.rs:38]: Yes, "fully-qualified sytax" is used for the form with the `<>`s. I would drop the extra clause, and just end after "functions."

File: library/alloc/src/rc.rs (hunk 2 above)
Comment (steveklabnik) [reply library/alloc/src/rc.rs:56]: We teach the `::clone` version in the book too; I wasn't notified of that thread either :)

Comment (camelid) [reply library/alloc/src/rc.rs:56]: I learned the `::clone` version (I guess from the book?) and I think it has clearer intent. Maybe for now we’ll go with it here and then we can change it later if we end up updating places like in the book?

File: library/alloc/src/rc.rs (hunk 1 above)
Comment (camelid) [reply library/alloc/src/rc.rs:38]: Hmm, wouldn’t it be better to give the readers somewhere to read more about fully qualified syntax? Side note, shouldn’t it be “fully-qualified syntax” with a hyphen?

**Comment by camelid** [comment]
Rebased.

**Comment by camelid** [comment]
Not again!

**Comment by camelid** [comment]
Rebased.

File: library/alloc/src/rc.rs (hunk 1 above)
Comment (steveklabnik) [reply library/alloc/src/rc.rs:38]: > wouldn’t it be better to give the readers somewhere to read more about fully qualified syntax?

Maybe. I am purely thinking about the wording itself.

> Side note, shouldn’t it be “fully-qualified syntax” with a hyphen?

I don't think so; the hyphen would be used if it is confusing which part "qualified" is attached to, that is, to disambiguate between

    fully-qualified syntax
    fully qualified-syntax

but I don't think that the latter would be a common interpretation

File: library/alloc/src/rc.rs (hunk 2 above)
Comment (cramertj) [reply library/alloc/src/rc.rs:56]: +1 for not using `::clone` as per https://github.com/rust-lang/rust/pull/63252
It's no longer the official guidance, and I don't think the existence of its use in other resources is good justification for continuing to use it in example code.

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 3]
```diff
@@ -1266,7 +1266,7 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         } else {
             err.span_suggestion(
                 span,
-                "use fully-qualified syntax",
+                "use fully qualified syntax",
```
Comment (cramertj) [inline compiler/rustc_typeck/src/astconv/mod.rs:1269]: Why was the hyphen here removed? It seems correct to me.

Comment (camelid) [reply compiler/rustc_typeck/src/astconv/mod.rs:1269]: I agree; I think it's correct with the hyphen. The issue is that it's referred to *without* the hyphen everywhere else in the compiler and in the [book](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#fully-qualified-syntax-for-disambiguation-calling-methods-with-the-same-name). I figure that unless and until we change it elsewhere, it should be consistent.

Comment (camelid) [reply compiler/rustc_typeck/src/astconv/mod.rs:1269]: See `@steveklabnik`'s comment: https://github.com/rust-lang/rust/pull/76138#discussion_r485706393.

File: library/alloc/src/rc.rs (hunk 2 above)
Comment (steveklabnik) [reply library/alloc/src/rc.rs:56]: still waiting for this change, @camelid . (I don't like it but it appears to be the consensus.)

**Review by steveklabnik — CHANGES_REQUESTED** [review]

**Comment by Dylan-DPC** [comment]
@camelid any updates on this? thanks

**Comment by camelid** [comment]
Yeah, I've been meaning to get to this.

**Comment by camelid** [comment]
Starting with a rebase.

**Comment by camelid** [comment]
I'm dropping the commit changing `fully-qualified syntax` to `fully qualified syntax` in the compiler since it should be a separate change.

**Comment by camelid** [comment]
@steveklabnik This is ready for a re-review :)

**Review by steveklabnik — APPROVED** [review]

**Comment by steveklabnik** [comment]
Thank you!

@bors: r+ rollup

**Pull request merged**